Add a glossary of Teleport terms

ptgott opened the pull request:
Closes #16121
Closes #13363

- Create a Core Concepts page based on the Architecture Overview page. The Overview page is an outline of core concepts, so I thought it would make a good starting point for the glossary.

- Add the Core Concepts page to the Home section to make it more prominent for users who are visiting the docs homepage. This should give it more traffic than leaving it in the Architecture section.

- Move definitions out of the "Definitions" table. Since this table repeats content from the body sections and comes after them, it makes sense to condense all content into brief body sections.

- Alphabetize sections

- Bold the first instance of each keyword within a given section. Not adding links since these are all to the same page. It might get a bit confusing to click a link and not see a change in view, since the target section would already be on the page.

- Remove the intro paragraph of the old Overview page, since it's not strictly related to core concepts.

ptgott committed:
Add a glossary of Teleport terms

Closes #16121
Closes #13363

- Create a Core Concepts page based on the Architecture Overview page.
  The Overview page is an outline of core concepts, so I thought it
  would make a good starting point for the glossary.

- Add the Core Concepts page to the Home section to make it more
  prominent for users who are visiting the docs homepage. This should
  give it more traffic than leaving it in the Architecture section.

- Move definitions out of the "Definitions" table. Since this table
  repeats content from the body sections and comes after them, it makes
  sense to condense all content into brief body sections.

- Alphabetize sections

- Bold the first instance of each keyword within a given section. Not
  adding links since these are all to the same page. It might get a bit
  confusing to click a link and not see a change in view, since the
  target section would already be on the page.

- Remove the intro paragraph of the old Overview page, since it's not
  strictly related to core concepts.

ptgott commented:
@xinding33 @klizhentas I'd like to block on a review from you since it's important that we have an approved set of terms that we can use throughout the docs as well as in the product. We should also make sure that the capitalization of Teleport-specific terms is correct.

ptgott commented:
**TODO:** I need to add a redirect to the Core Concepts page

xinding33 commented on docs/pages/core-concepts.mdx:
```diff
@@ -0,0 +1,219 @@
+
+Read more in our [Resource Reference](./reference/resources.mdx).
+
+## Machine ID
```
It's worth mentioning here that Machine ID is commonly used for machine-to-machine and service-to-service use cases.

xinding33 commented on docs/pages/core-concepts.mdx:
```diff
@@ -0,0 +1,219 @@
+Proxy Service. In a demo environment, you can run these two services from a
+single **Teleport Process**.
+
+## Teleport Control Plane
+
+The Control Plane consists of the **Auth Service** and **Proxy Service**, which
+govern cluster-wide access. Other **Teleport Services**, which ensure access to
+specific resources, are not considered part of the Teleport Control Plane.
```
Let's deprecate the "Teleport Control Plane" terminology altogether in favor or "Teleport Cluster."

xinding33 commented on docs/pages/core-concepts.mdx:
```diff
@@ -0,0 +1,219 @@
+The key concept of Teleport's architecture is the **cluster**. A Teleport
+cluster consists of the **Teleport Auth Service** and **Teleport Proxy Service**,
+plus several **Teleport Services** that manage traffic to resources within your
+infrastructure, such as Kubernetes clusters and Windows desktops. A Teleport
+cluster also includes **Machine ID** and **Teleport Discover**.
```
> A Teleport cluster also includes **Machine ID** and **Teleport Discover**

This part didn't make sense to me. I would clarify that Machine ID and Teleport Discover are both services you can run in a cluster.

zmb3 replied:
Teleport Discover isn't really a service that you run in the cluster though - is it?

I'm not sure we should even mention discover here. That feels more like an internal name to me. It's not used anywhere in the UI (or docs, to my knowledge)

ptgott replied:
[This PR](https://github.com/gravitational/teleport/pull/16279) introduces the "Teleport Discovery Service", which maybe we could mention instead of Teleport Discover?

xinding33 commented on docs/pages/core-concepts.mdx:
```diff
@@ -0,0 +1,219 @@
+these terms often when setting up and managing Teleport, so you should get
+familiar with them before following other pages in the documentation.
+
+## Authentication connector
```
I think it would improve readability and comprehension if we pulled out the primary terminologies into its own section at the top and then have an alphabetical list of terminologies in a secondary section. For example, we should pull Teleport Cluster, Teleport Process, and Teleport Service out and put them at the top so users understand the fundamental pieces of Teleport before reading further.

xinding33 commented on docs/pages/core-concepts.mdx:
```diff
@@ -0,0 +1,219 @@
+govern cluster-wide access. Other **Teleport Services**, which ensure access to
+specific resources, are not considered part of the Teleport Control Plane.
+
+## Teleport Enterprise
+
+**Teleport Enterprise** is a paid plan that includes additional features for the
+**Teleport Control Plane** and **Teleport Services**, plus a support agreement
+with Teleport.
+
+Read more in our [Teleport Enterprise
+guide](./deploy-a-cluster/teleport-enterprise/introduction.mdx).
```
I'm not a fan of the terminology "Teleport Enterprise" because it implies that Teleport Cloud is not enterprise. This is how I think about our different offerings:

- Teleport Community Edition (this is open source, free, and self-hosted only)
- Teleport Enterprise Edition (this is paid and can be self-hosted or SaaS)
    - Teleport Enterprise Self-Hosted
    - Teleport Enterprise Cloud

ptgott commented:
Hi @klizhentas, I am just following up to request a review. I was hoping to get your feedback before responding to Xin's so I can incorporate them together. Thanks!

zmb3 commented on docs/pages/core-concepts.mdx:
```diff
@@ -0,0 +1,219 @@
+|[Teleport Application Service](./application-access/introduction.mdx)|HTTP and TCP traffic to user-configured endpoints, e.g., internal web applications or the AWS Console.|
+|[Teleport Desktop Service](./desktop-access/introduction.mdx)|Remote Desktop Protocol traffic to Windows desktops.|
+|[Teleport Kubernetes Service](./kubernetes-access/introduction.mdx)|HTTP traffic to the Kubernetes API server.|
+|[Teleport Node Service](./server-access/introduction.mdx)|Remote shell command execution via an SSH server implementation. Machines you can access via this service are called **Teleport Nodes**.|
```
Let's call this SSH service, since that's how it's referred to when configuring teleport.

ptgott committed:
Respond to PR feedback and add further edits

- Group definitions into H3-level sections organized into H2s, improving
  readability by presenting the more central components first.

- Group "Machine ID" with "Teleport Services" to simplify the glossary.

- Call instances of a Teleport Service an "agent". This is consistent
  with our source and informal language about Teleport. This also
  removes the "Node" terminology, which is prone to misuse (does it mean
  "a Service instance"? An "SSH Service instance"? Both are present in
  the docs) and easily confused with the lowercase "node" (i.e., a
  host).

- For simplicity, remove the term "Teleport Process" and simply explain
  that multiple Teleport Services can run via the same binary.

- Add a section on Teleport's editions, including the terms "Community
  Edition", "Enterprise", and "Enterprise Cloud".

- Remove the mention of Teleport Discover, which isn't documented yet.

- Minor wording/readability changes.

ptgott commented:
@xinding33 @klizhentas I have pushed a second draft of the glossary.

alexfornuto commented on docs/pages/core-concepts.mdx:
```diff
@@ -0,0 +1,245 @@
+description: Learn the the key architectural components that make up Teleport.
+---
+
+Here are the core concepts describe a Teleport deployment. You will see these
```
```suggestion
Here are the core concepts that describe a Teleport deployment. You will see these
```

alexfornuto commented on docs/pages/core-concepts.mdx:
```diff
@@ -0,0 +1,245 @@
+
+The key concept of Teleport's architecture is the **cluster**. A Teleport
+cluster consists of the **Teleport Auth Service** and **Teleport Proxy Service**,
+plus several **Teleport Services** that manage traffic to resources within your
```
```suggestion
plus one or more **Teleport Services** that manage traffic to resources within your
```

ptgott replied:
Suggesting using "the", since our docs sometimes imply that a Teleport cluster consists only of the Auth/Proxy Service

alexfornuto commented on docs/pages/core-concepts.mdx:
```diff
@@ -0,0 +1,245 @@
+### Agent
+
+An instance of a **Teleport Service** is called an **agent**. For all Teleport
+Services besides the **Teleport SSH Service**, an agent  and can enable access
```
```suggestion
Services besides the **Teleport SSH Service**, an agent can enable access
```

alexfornuto reviewed (approved):
👍 with some nits.

ptgott commented on docs/pages/core-concepts.mdx:
```diff
@@ -0,0 +1,245 @@
+  caption="Auth service"
+>
+
+![Auth service](../img/architecture/auth.png)
```
I used the images in our original Architectural Overview page here. So we can get this out more quickly, I've suggested removing the image references from this page—maybe we can create a separate issue to fix the images and potentially rework the use of diagrams in the Core Concepts page?

ptgott committed:
Respond to PR feedback

ptgott commented:
@xinding33 When you have a moment, would you mind taking another look at this to make sure it provides the changes you requested?

@klizhentas I would also appreciate a review from you, if you have time, since this page touches on the fundamental ways we talk/write about Teleport.

Thanks!

zmb3 commented on docs/pages/core-concepts.mdx:
```diff
@@ -0,0 +1,224 @@
+
+A minimal Teleport cluster consists of the **Teleport Auth Service** and
+**Teleport Proxy Service**. In a demo environment, you can run these two
+services from a single `teleport` binary on a [Linux
```
```suggestion
services from a single `teleport` process on a [Linux
```

zmb3 commented on docs/pages/core-concepts.mdx:
```diff
@@ -0,0 +1,224 @@
+---
+title: Teleport Core Concepts
+description: Learn the the key architectural components that make up Teleport.
```
```suggestion
description: Learn the the key components that make up Teleport.
```

zmb3 commented on docs/pages/core-concepts.mdx:
```diff
@@ -0,0 +1,224 @@
+issues certificates to clients and maintains an audit log.
+
+The Auth Service is the only component of a cluster that has to be connected to
+a backend, which it uses to store **configuration resources** and the
```
```suggestion
a backend, which it uses to store **cluster state ** and the
```

zmb3 commented on docs/pages/core-concepts.mdx:
```diff
@@ -0,0 +1,224 @@
+as Kubernetes clusters, Windows desktops, internal web applications, and
+databases.
+
+A single running `teleport` binary can run one or more **Teleport Services**,
```
```suggestion
A single running `teleport` process can run one or more **Teleport Services**,
```

zmb3 commented on docs/pages/core-concepts.mdx:
```diff
@@ -0,0 +1,224 @@
+
+Read more in our [Machine ID guide](./machine-id/introduction.mdx).
+
+### Agent
```
With the push towards agentless are we sure we want to formally define agent here?

ptgott replied:
I was looking for a term we could use to describe an instance of a Teleport service. We already use "agent" in the docs, the Teleport source, and conversations with users/prospects, so this seemed appropriate. I know there are misconceptions around the use of agents in Teleport (https://github.com/gravitational/teleport/issues/13351), but I thought we could add clarity to the term here rather than choose a different term.

I also think the term "agent" is better than alternatives. While we could reappropriate "Node" to mean "any service that manages access to a resource", I think the "Node/node" distinction is unwieldy. We could say "Service instance", but it has twice the words/syllables. There's also "relay", as per StrongDM, but that doesn't really describe the SSH Service.

What do you think?

zmb3 replied:
I don't feel strongly about this, totally up to you. Agent is fine with me.

zmb3 commented on docs/pages/core-concepts.mdx:
```diff
@@ -0,0 +1,224 @@
+  e.g., Okta or GitHub. When a user authenticates to Teleport via your SSO
+  solution, Teleport issues a certificate for the user and creates a temporary
+  **local user** that is valid for the lifetime of the certificate.
+- **Bot users** are automated services that receive credentials via **Machine ID**.
```
Technically there isn't actually a resource that is called bot user. They are the same as local users, they just don't have passwords.

ptgott committed:
Respond to zmb3 feedback
